lib/server.js

```diff
@@ -826,12 +825,32 @@ Server.prototype._run = function _run(req, res, route, chain, cb) {
 
         if (arg) {
             if (arg instanceof Error) {
-                errName = arg.name.replace(/Error$/, '');
-                log.trace({err: arg, errName: errName}, 'next(err=%s)',
-                    (arg.name || 'Error'));
 
+                var errName = arg.name.replace(/Error$/, '');
+                log.trace({
+                    err: arg,
+                    errName: errName
+                }, 'next(err=%s)', (arg.name || 'Error'));
+
+                // always attempt to use the most specific error listener
+                // possible. fall back on generic 'error' listener if we can't
+                // find one for the error we got.
+                var hasErrListeners = false;
+                var errEvtName;
+
+                // if we have listeners for the specific error
                 if (self.listeners(errName).length > 0) {
-                    self.emit(errName, req, res, arg, once(function () {
+                    hasErrListeners = true;
+                    errEvtName = errName;
+                }
+                // or if we have a generic error listener
+                else if (self.listeners('error').length > 0) {
+                    hasErrListeners = true;
+                    errEvtName = 'error';
```

Comment:
restifyError sounds good. +1 on updating docs to better reflect the
different classes of errors.

On 14 September 2015 at 18:35, Alex Liu <notifications@github.com> wrote:

> In lib/server.js
> <https://github.com/restify/node-restify/pull/892#discussion_r39466698>:
>
> >                  if (self.listeners(errName).length > 0) {
> > -                    self.emit(errName, req, res, arg, once(function () {
> > +                    hasErrListeners = true;
> > +                    errEvtName = errName;
> > +                }
> > +                // or if we have a generic error listener
> > +                else if (self.listeners('error').length > 0) {
> > +                    hasErrListeners = true;
> > +                    errEvtName = 'error';
>
> I went with 'nextError', but happy to change it to 'restifyError'. I
> thought it might reduce the number of questions we get around the
> difference between InternalError vs InternalServerError vs
> uncaughtException, but maybe that's a docs issue.
>
> —
> Reply to this email directly or view it on GitHub
> <https://github.com/restify/node-restify/pull/892/files#r39466698>.
>